geomstats/geometry/functions.py

```diff
@@ -0,0 +1,245 @@
+"""Module for function spaces as geometric objects"""
+
+import math
+
+import numpy as np
+
+import geomstats.backend as gs
+from geomstats.geometry.manifold import Manifold
+from geomstats.geometry.riemannian_metric import RiemannianMetric
+
+
+class SinfSpaceMetric(RiemannianMetric):
+    r"""A Riemannian metric on the S_{\inf} space
+
+    Parameters:
+    -------
+    domain_samples : array of shape (n_samples, )
+        grid points on the domain
+    """
+
+    def __init__(self, domain_samples):
+        self.domain = domain_samples
+        self.x = (self.domain - min(self.domain)) / (
+            max(self.domain) - min(self.domain)
+        )
+        self.n_evals = len(self.domain)
+        super().__init__(dim=self.n_evals)
+
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point=None):
+        """Inner product between two tangent vectors at a base point.
+
+        Parameters
+        ----------
+        tangent_vec_a: array-like, shape=[..., n_evals]
+            Tangent vector at base point.
+        tangent_vec_b: array-like, shape=[..., n_evals]
+            Tangent vector at base point.
+        base_point: array-like, shape=[..., n_evals]
+            Base point.
+            Optional, default: None.
+
+        Returns
+        -------
+        inner_product : array-like, shape=[...,]
+            Inner-product.
+        """
+
+        def _f(v1, v2):
+            return gs.cast(np.trapz(v1 * v2, x=self.x), gs.float32)
+
+        inner_prod = gs.vectorize(
+            (tangent_vec_a, tangent_vec_b),
+            _f,
+            dtype=gs.float32,
+            multiple_args=True,
+            signature="(i),(i)->()",
+        )
+
+        return inner_prod
+
+    def exp(self, tangent_vec, base_point, **kwargs):
+        """Compute the Riemannian exponential of a tangent vector.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., n_evals]
+            Tangent vector at a base point.
+        base_point : array-like, shape=[..., n_evals]
+            Point on the hypersphere.
+
+        Returns
+        -------
+        exp : array-like, shape=[..., n_evals]
+            Point on the hypersphere equal to the Riemannian exponential
+            of tangent_vec at the base point.
+        """
+
+        def _f(v, p):
+            norm_v = self.norm(v)
+            t1 = gs.cos(norm_v) * p
+            t2 = (gs.sin(norm_v) / norm_v) * p
```

Comment:
All these operations are already vectorized or can be by using einsum, please have a look at the finite dim equivalent. Also the use of taylor_exp_even_func make sure we stay precise around 0.
https://github.com/geomstats/geomstats/blob/d89ee0a4eb8cd178a5de5bccc095fda52d9c0732/geomstats/geometry/hypersphere.py#L743-L747

Reply:
Great, this is really helpful. I tried to use as much code as possible but it was not straightforward to define the new metric using the `HypersphereMetric`. The main challenge is we need the embedding metric to be  `L2Metric` on functions. I did not want to define one separately but if the existing `L2Metric` can be updated (somehow), I think I can write the new metric mostly as another sub-class of Hypersphere metric.